Foundation/NSDecimal.swift

```diff
@@ -1,155 +1,1837 @@
+                q -= 1
+                rtmp += UInt32(v1)
+            }
+        }
+
+        // D4: multiply and subtract.
+
+        var mk:UInt32 = 0 // multiply carry
+        var sk:UInt32 = 1 // subtraction carry
+        var acc:UInt32
+
+        // We perform a multiplication and a subtraction
+        // during the same pass...
+        for i in 0...v._length {
+            let ul = u._length
+            let vl = v._length
+            acc = q * UInt32(v[i]) + mk     // multiply
+            mk = acc >> 16                  // multiplication carry
+            acc = acc & 0xffff;
+            let uu = u // work around compiler bug
+            acc = 0xffff + UInt32(uu[ul - vl + i - UInt32(j) - UInt32(1)]) - acc + sk; // subtract
+            sk = acc >> 16;
+            u[ul - vl + i - UInt32(j) - UInt32(1)] = UInt16(truncatingBitPattern:acc)
+        }
+
+        // D5: test remainder
+        // This test catches cases where q is still q + 1
+        if sk == 0 {
+            // D6: add back.
+            var k:UInt32 = 0 // Addition carry
+
+            // subtract one from the quotient digit
+            q -= 1
+            for i in 0...v._length {
+                let ul = u._length
+                let vl = v._length
+                let vv = v // work around compiler bug
+                let uu = u // work around compiler bug
+                acc = UInt32(vv[i]) + UInt32(uu[UInt32(ul) - UInt32(vl) + UInt32(i) - UInt32(j) - UInt32(1)]) + k
+                k = acc >> 16;
+                u[UInt32(ul) - UInt32(vl) + UInt32(i) - UInt32(j) - UInt32(1)] = UInt16(truncatingBitPattern:acc)
+            }
+            // k must be == 1 here
+        }
+
+        r[UInt32(ql - j - UInt16(1))] = UInt16(q)
+        // D7: loop on j
+    }
 
-public func NSDecimalRound(_ result: UnsafeMutablePointer<Decimal>, _ number: UnsafePointer<Decimal>, _ scale: Int, _ roundingMode: NSDecimalNumber.RoundingMode) { NSUnimplemented() }
+    r._length = UInt32(ql);
+
+    while r._length != 0 && r[r._length - 1] == 0 {
+        r._length -= 1
+    }
+
+    return .noError;
+}
+
+fileprivate func integerMultiplyByPowerOf10(_ result: inout WideDecimal, _ left: WideDecimal, _ p: Int) -> NSDecimalNumber.CalculationError {
+    var power = p
+    if power == 0 {
+        result = left
+        return .noError
+    }
+    let isNegative = power < 0
+    if isNegative {
+        power = -power
+    }
+    result = left
+
+    let maxpow10 = pow10.count - 1
+    var error:NSDecimalNumber.CalculationError = .noError
+
+    while power > maxpow10 {
+        var big = WideDecimal()
+
+        power -= maxpow10
+        let p10 = WideDecimal(pow10[maxpow10])
+
+        if !isNegative {
+            error = integerMultiply(&big,result,p10)
+        } else {
+            error = integerDivide(&big,result,p10)
+        }
+
+        if error != .noError && error != .lossOfPrecision {
+            return error;
+        }
+
+        for i in 0..<Decimal.maxSize {
```

Comment:
This should probably be big._length rather than Decimal.maxSize, otherwise we'll lose precision